### PR TITLE
[SPARK-15672][R][DOC] R programming guide update

### PR DESCRIPTION
**Pull request by vectorijk**
## What changes were proposed in this pull request?

Guide for
- UDFs with dapply, dapplyCollect
- spark.lapply for running parallel R functions
## How was this patch tested?

build locally
screen shot 2016-06-14 at 03 12 56

### CODE REVIEW

**Comment by vectorijk** [comment]
cc @jkbradley @shivaram

File: docs/sparkr.md [hunk 1]
```diff
@@ -262,6 +262,67 @@ head(df)
 {% endhighlight %}
 </div>
 
+### Applying User-defined Function
+
```
Comment (shivaram) [inline docs/sparkr.md]: It will be good to add an introduction here that there are two kinds of user-defined functions we support in SparkR. Something like 
```
In SparkR we support two kinds for user-defined functions
1. Run a given function on a large dataset using dapply. 
2. Run many functions in parallel using spark.lapply. 
```

File: docs/sparkr.md [hunk 2]
```diff
@@ -262,6 +262,67 @@ head(df)
...
+#### dapply
+Apply a function to each partition of `SparkDataFrame`. The function to be applied to each partition of the `SparkDataFrame` and should have only one parameter, to which a `data.frame` corresponds to each partition will be passed. The output of function should be a `data.frame`.
+<div data-lang="r"  markdown="1">
+{% highlight r %}
+
+# Convert waiting time from hours to seconds.
+# Note that we can apply UDF to DataFrame.
+
+df1 <- dapply(df, function(x) {x}, schema(df))
```
Comment (shivaram) [inline docs/sparkr.md]: The conversion is not actually happening in this example ?

File: docs/sparkr.md [hunk 3]
```diff
@@ -262,6 +262,67 @@ head(df)
...
+{% highlight r %}
+
+# Convert waiting time from hours to seconds.
+# Note that we can apply UDF to DataFrame.
+ldf <- dapplyCollect(
+         df,
+         function(x) {
+           x <- cbind(x, "waiting_secs"=x$waiting * 60)
+         })
+head(df, 3)
```
Comment (shivaram) [inline docs/sparkr.md]: Would be good to show how `head(Ldf)` looks here. Also we should note that the difference in dapplyCollect is that the schema doesn't need to be passed in by the user

**Comment by shivaram** [comment]
Thanks @vectorijk - I left some comments inline.

cc @felixcheung

File: docs/sparkr.md (hunk 2 above)
Comment (vectorijk) [inline docs/sparkr.md]: ok, I will improve this example more specifically.

File: docs/sparkr.md [hunk 4]
```diff
@@ -262,6 +262,67 @@ head(df)
 {% endhighlight %}
 </div>
 
+### Applying User-defined Function
+
+#### dapply
+Apply a function to each partition of `SparkDataFrame`. The function to be applied to each partition of the `SparkDataFrame` and should have only one parameter, to which a `data.frame` corresponds to each partition will be passed. The output of function should be a `data.frame`.
```
Comment (felixcheung) [inline docs/sparkr.md]: perhaps explain why the schema needs to be passed here?

File: docs/sparkr.md [hunk 5]
```diff
@@ -262,6 +262,67 @@ head(df)
...
+<div data-lang="r"  markdown="1">
+{% highlight r %}
+
+# Perform distributed training of multiple models with spark.lapply
+families <- c("gaussian", "poisson")
+train <- function(family) {
+  model <- glm(Sepal.Length ~ Sepal.Width + Species, iris, family = family)
+  summary(model)
+}
+model.summaries <- spark.lapply(sc, families, train)
```
Comment (felixcheung) [inline docs/sparkr.md]: perhaps describe more on what will get passed to this udf `train` here?

File: docs/sparkr.md (hunk 3 above)
Comment (felixcheung) [inline docs/sparkr.md]: it will also be useful to point out `ldf` is a `data.frame`?

**Comment by jkbradley** [comment]
Ping @vectorijk

**Comment by vectorijk** [comment]
@jkbradley @shivaram @felixcheung addressed comments.

**Comment by vectorijk** [comment]
Jenkins test this again.

**Comment by felixcheung** [comment]
great! please see pending PR #13752 on removing `sc` parameter from `spark.lapply`

**Comment by NarineK** [comment]
Hi @vectorijk , @felixcheung ,
As I was looking at the documentation generated in R I've noticed that there is some duplicated information. I'm not sure if this is the right place to ask about it, I thought you might have seen it.
In I help I see the following:
```
Arguments

x	
A SparkDataFrame
func	
A function to be applied to each partition of the SparkDataFrame. func should have only one parameter, to which a data.frame corresponds to each partition will be passed. The output of func should be a data.frame.
schema	
The schema of the resulting SparkDataFrame after the function is applied. It must match the output of func.
x	
A SparkDataFrame
func	
A function to be applied to each partition of the SparkDataFrame. func should have only one parameter, to which a data.frame corresponds to each partition will be passed. The output of func should be a data.frame.
See Also

Other SparkDataFrame functions: SparkDataFrame-class, [[, agg, arrange, as.data.frame, attach, cache, collect, colnames, coltypes, columns, count, createOrReplaceTempView, describe, dim, distinct, dropDuplicates, dropna, drop, dtypes, except, explain, filter, first, gapplyCollect, gapply, group_by, head, histogram, insertInto, intersect, isLocal, join, limit, merge, mutate, ncol, persist, printSchema, rename, repartition, sample, saveAsTable, selectExpr, select, showDF, show, str, take, unionAll, unpersist, withColumn, with, write.df, write.jdbc, write.json, write.parquet, write.text

Other SparkDataFrame functions: SparkDataFrame-class, [[, agg, arrange, as.data.frame, attach, cache, collect, colnames, coltypes, columns, count, createOrReplaceTempView, describe, dim, distinct, dropDuplicates, dropna, drop, dtypes, except, explain, filter, first, gapplyCollect, gapply, group_by, head, histogram, insertInto, intersect, isLocal, join, limit, merge, mutate, ncol, persist, printSchema, rename, repartition, sample, saveAsTable, selectExpr, select, showDF, show, str, take, unionAll, unpersist, withColumn, with, write.df, write.jdbc, write.json, write.parquet, write.text

```

Is this on purpose ?

**Comment by felixcheung** [comment]
@NarineK That is sort of unrelated to this PR since this PR is about the programming guide?

But in short, this happens because in the R code both `dapply` and `dapplyCollect` has the `@rdname` tag to "dapply". I'm not sure if we need to do that. But the first copy of "x ..." and "func ..." is from "dapply" and the second is from "dapplyCollect".

**Comment by shivaram** [comment]
Yeah we can remove the duplication by having separate rd files or by just removing documentation for the overlapping arguments (I think in this case `x` and `func` are the same for `dapply` and `dapplyCollect`). 

@NarineK feel free to open a separate JIRA/PR for this

File: docs/sparkr.md [hunk 6]
```diff
@@ -262,6 +262,79 @@ head(df)
...
+##2     1.800      54         3240
+##3     3.333      74         4440
+##4     2.283      62         3720
+##5     4.533      85         5100
+##6     2.883      55         3300
+{% endhighlight %}
+</div>
+
+##### dapplyCollect
+Like `dapply`, apply a function to each partition of `SparkDataFrame` and collect the result back.
```
Comment (shivaram) [inline docs/sparkr.md]: I think its good to say a couple of things here. First that we don't require any schema to be passed in to `dapplyCollect` (unlike `dapply`). The other thing is that its good to remind users that this should be used only if the output of the UDF run on all the partitions can fit in driver memory.

File: docs/sparkr.md [hunk 7]
```diff
@@ -262,6 +262,79 @@ head(df)
...
+##3     3.333      74         4440
+
+{% endhighlight %}
+</div>
+
+#### Run many functions in parallel using `spark.lapply`
+
+##### lapply
+Similar to `lapply` in native R, `spark.lapply` runs a function over a list of elements and distributes the computations with Spark.
+Applies a function in a manner that is similar to `doParallel` or `lapply` to elements of a list.
```
Comment (shivaram) [inline docs/sparkr.md]: Similar to the above, it would be good to add a line here saying the results of all the computations should fit in a single machine -- And that if that is not the case they can do something like `df <- createDataFrame(list)` and then use `dapply`

Comment (vectorijk) [inline docs/sparkr.md]: Thanks so much for pointing out this! I will update those very soon.

**Comment by sun-rui** [comment]
Can you add documentation for gapply() and gapplyCollect() together here? or @NarineK will do in another PR?

File: docs/sparkr.md [hunk 8]
```diff
@@ -262,6 +262,83 @@ head(df)
 {% endhighlight %}
 </div>
 
+### Applying User-defined Function
+In SparkR, we support several kinds for User-defined Functions:
```
Comment (sun-rui) [inline docs/sparkr.md]: several kinds of?

File: docs/sparkr.md [hunk 9]
```diff
@@ -262,6 +262,83 @@ head(df)
...
+
+#### Run a given function on a large dataset using `dapply` or `dapplyCollect`
+
+##### dapply
+Apply a function to each partition of `SparkDataFrame`. The function to be applied to each partition of the `SparkDataFrame`
```
Comment (sun-rui) [inline docs/sparkr.md]: Apply a function to each partition of a `SparkDataFrame`.

File: docs/sparkr.md [hunk 10]
```diff
@@ -262,6 +262,83 @@ head(df)
...
+##2     1.800      54         3240
+##3     3.333      74         4440
+##4     2.283      62         3720
+##5     4.533      85         5100
+##6     2.883      55         3300
+{% endhighlight %}
+</div>
+
+##### dapplyCollect
+Like `dapply`, apply a function to each partition of `SparkDataFrame` and collect the result back. The output of function
```
Comment (sun-rui) [inline docs/sparkr.md]: apply a function to each partition of a `SparkDataFrame`

File: docs/sparkr.md [hunk 11]
```diff
@@ -262,6 +262,83 @@ head(df)
...
+output of UDF run on all the partitions can fit in driver memory.
+<div data-lang="r"  markdown="1">
+{% highlight r %}
+
+# Convert waiting time from hours to seconds.
+# Note that we can apply UDF to DataFrame and return a R's data.frame
+ldf <- dapplyCollect(
+         df,
+         function(x) {
+           x <- cbind(x, "waiting_secs"=x$waiting * 60)
```
Comment (sun-rui) [inline docs/sparkr.md]: style nit: "waiting_secs" = x$waiting * 60

File: docs/sparkr.md [hunk 12]
```diff
@@ -262,6 +262,83 @@ head(df)
...
+##1     3.600      79         4740
+##2     1.800      54         3240
+##3     3.333      74         4440
+
+{% endhighlight %}
+</div>
+
+#### Run many functions in parallel using `spark.lapply`
+
+##### lapply
```
Comment (sun-rui) [inline docs/sparkr.md]: spark.lapply

Comment (sun-rui) [inline docs/sparkr.md]: One thought about spark.lapply() is that documenting here means our commitment to it. This is a case demonstrating the need to support Dataset in SparkR. Maybe next step we can consider replace RDD with Dataset in SparkR

Comment (shivaram) [inline docs/sparkr.md]: This dicussion probably belongs on JIRA, but I don't think spark.lapply is particularly similar to RDD or Dataset. We are taking a local list here and passing a value from that to each parallel function -- I think of Datasets / RDDs as distributed datasets which don't fit in local R session and DataFrames in SparkR already cover that.

Comment (felixcheung) [inline docs/sparkr.md]: I think spark.lapply is more about running native R distributed vs Dataset would be typed data distributed (without native R), that seems fairly orthogonal to me.

speaking of, how should we address local R here? should this say "Run local R functions in parallel using spark.lapply" or "Run local R functions distributed using spark.lapply"?

Comment (shivaram) [inline docs/sparkr.md]: I'm a little more inclined to use `distributed` instead of `parallel` but both of them sound fine to me.

File: docs/sparkr.md (hunk 9 above)
Comment (NarineK) [inline docs/sparkr.md]: @sun-rui , will do gapply in another PR.

Comment (vectorijk) [inline docs/sparkr.md]: @NarineK I filed a JIRA [SPARK-16112](https://issues.apache.org/jira/browse/SPARK-16112) for gapply programming guide so that you could open a PR for this.

**Comment by shivaram** [comment]
@felixcheung @jkbradley any more comments on this ?

**Comment by jkbradley** [comment]
LGTM.  I'll merge this with master and branch-2.0
Thanks!
